### PR TITLE
Reintroduce hyperparameter tuning

### LINKED ISSUE #2632: Reintroduce hyperparameter tuning in Flair
Hyperparameter tuning used to be a feature in Flair<0.10. There's still some remains of this feature in [the documentation](https://github.com/flairNLP/flair/blob/master/resources/docs/KOR_docs/TUTORIAL_8_MODEL_OPTIMIZATION.md) though the code for it has been removed.

My understanding is that the feature was removed because noone was prepared to maintain it. Some users are still asking about why it was removed and are probably surprised to see it go: https://github.com/flairNLP/flair/pull/2518

The work required to get hyperparameter tuning working again with changes from Flair v0.10 is pretty minor.
I'd be happy to reintroduce hyperparameter tuning in Flair in its basic form.

### PR DESCRIPTION
**Pull request by tadejmagajna**
This PR reintroduces hyperparameter tuning in Flair
closes #2632

A number of small changes was required so I split this PR into appropriately named commits where each commit is an isolated change to the code base. For example, there is a separate commit that reverts the deletion or hyperparameter tuning in Flair - this helps us quickly see which code changes are new and which simply reintroduce the old code.

The changes and their corresponding commits include:
1. [Revert "Removes hyperparameter features"](https://github.com/flairNLP/flair/commit/c9591b9482afcd3971497fe88afb14bc394bd516)[](https://github.com/tadejmagajna)  Reverts the removal of all hyperparameter runing code
2. [Updating the param selection docs for the v0.10 syntax](https://github.com/flairNLP/flair/commit/32ebc4c13ddfcebf60a0c8487000d1359a497e86). This updates the README tutorial for English and Korean (I did my best with Korean here) to fit the new v0.10 compliant syntax.
3. [Adding hyperopt back to requirements.txt](https://github.com/flairNLP/flair/commit/f0c9f42a31468b4ee76497dd18a9d02f277b3479) adds the most recent version of hyperopt back to requirements
4. [Fixing paramselection code to work with changes in Flair v0.10](https://github.com/flairNLP/flair/commit/2995cc43153516f9709f58c33c449899dd7bf5f1) This updates the original hyperparameter syntax to work with changes applied in v0.10 such as, for example, the `make_label_dictionary(label_type)` requiring the `label_type` argument
5. [Fixing bug where embeddings got added twice on multiple training runs](https://github.com/flairNLP/flair/commit/0a140b8a7803c44228b4943a22dab3a67f6d87b8)[](https://github.com/tadejmagajna). This provides a temporary solution for #2600 where embeddings were added to tokens multiple times and this caused crashes (can be reproduced by running the TUTORIAL_8 code). Note that the fix probably does probably slow the execution down a bit, but it only affects runs where `training_runs>1`
6. [Enabling and fixing tests for param selection](https://github.com/flairNLP/flair/commit/a63f988a5db2c8471947d24f77cd302f803b5ed6). Here I reintroduce hyperparameter tests, apply some fixes to them and enable them again because they were previously disabled for some reason. There was also a deprecation warning related to some dependency of Hyperopt. To fix it I applied a change to to `pyproject.toml` where I swapped the `-W` parameter for the more-readable `filterwarnings` parameter and made it ignore that specific `DeprecationWarning`.
7. [Automatic code formatting of param selection tests](https://github.com/flairNLP/flair/commit/ebaa9fe4ca37f91964bc84c5c7244ba1a41c06e7) auto code formatting

Please do me know if there are any other parts of the hyperparameter tuning code I forgot to make compliant with the current Flair syntax

### CODE REVIEW

File: flair/hyperparameter/param_selection.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        self.multi_label = multi_label
+        self.label_type = label_type
+        self.document_embedding_type = document_embedding_type
+
+        self.label_dictionary = self.corpus.make_label_dictionary(self.label_type)
+
+    def _set_up_model(self, params: dict):
+        embdding_params = {key: params[key] for key in params if key in DOCUMENT_EMBEDDING_PARAMETERS}
+
+        if self.document_embedding_type == "lstm":
```
Comment (alanakbik) [inline flair/hyperparameter/param_selection.py:248]: it seems the only options are RNN or pooled word embeddings, but the standard approach today would be to simply use a transformer only which far outperforms RNNs trained over word embeddings. Not necessarily something that needs to be fixed in this PR but something that limits the usefulness of our parameter selection for text classifiers at the moment.

File: flair/hyperparameter/parameter.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    Parameter.HIDDEN_SIZE.value,
+    Parameter.RNN_LAYERS.value,
+    Parameter.USE_CRF.value,
+    Parameter.USE_RNN.value,
+    Parameter.DROPOUT.value,
+    Parameter.LOCKED_DROPOUT.value,
+    Parameter.WORD_DROPOUT.value,
+]
+DOCUMENT_EMBEDDING_PARAMETERS = [
+    Parameter.EMBEDDINGS.value,
```
Comment (alanakbik) [inline flair/hyperparameter/parameter.py:57]: one idea for the problem above would be to support param selection only for `TransformerDocumentEmbeddings` and define some parameters that one would like to search over (which transformer model, which layers, fine-tune yes/no) and remove the RNN specific ones.

File: resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+[hyperopt](https://github.com/hyperopt/hyperopt).
+
+First you need to load your corpus. If you want to load the [AGNews corpus](https://www.di.unipi.it/~gulli/AG_corpus_of_news_articles.html)
+used in the following example, you first need to download it and convert it into the correct format. Please
+check [tutorial 6](/resources/docs/TUTORIAL_6_CORPUS.md) for more details.
+```python
+from flair.datasets import TREC_6
+
+# load your corpus
+corpus = TREC_6()
```
Comment (alanakbik) [inline resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md:18]: The tutorial is a bit counterintuitive since it defines a text classification corpus and then word embeddings. The fact that there's an LSTM used is buried in the model initialization and not explained.

Perhaps switch to a sequence labeling example? `WNUT_17` for instance could be used as example corpus, and for sequence labeling the current implementation allows some interesting parameter searches.

File: resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+# 1. get the corpus
+corpus = WNUT_17().downsample(0.1)
+print(corpus)
+
+# 2. what tag do we want to predict?
+tag_type = 'ner'
+
+# 3. make the tag dictionary from the corpus
+tag_dictionary = corpus.make_tag_dictionary(tag_type=tag_type)
```
Comment (alanakbik) [inline resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md:121]: should be `tag_dictionary = corpus.make_label_dictionary(label_type=tag_type)`, otherwise the example won't run

File: resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                                        embeddings=embeddings,
+                                        tag_dictionary=tag_dictionary,
+                                        tag_type=tag_type,
+                                        use_crf=True)
+
+# 6. initialize trainer
+trainer: ModelTrainer = ModelTrainer(tagger, corpus)
+
+# 7. find learning rate
+learning_rate_tsv = trainer.find_learning_rate('resources/taggers/example-ner', Adam)
```
Comment (alanakbik) [inline resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md:144]: this function is broken, it should be changed to the code below, ideally as part of this PR.

```python
    def find_learning_rate(
        self,
        base_path: Union[Path, str],
        optimizer,
        file_name: str = "learning_rate.tsv",
        start_learning_rate: float = 1e-7,
        end_learning_rate: float = 10,
        iterations: int = 100,
        mini_batch_size: int = 32,
        stop_early: bool = True,
        smoothing_factor: float = 0.98,
        **kwargs,
    ) -> Path:
        best_loss = None
        moving_avg_loss = 0

        # cast string to Path
        if type(base_path) is str:
            base_path = Path(base_path)
        learning_rate_tsv = init_output_file(base_path, file_name)

        with open(learning_rate_tsv, "a") as f:
            f.write("ITERATION\tTIMESTAMP\tLEARNING_RATE\tTRAIN_LOSS\n")

        optimizer = optimizer(self.model.parameters(), lr=start_learning_rate, **kwargs)

        train_data = self.corpus.train

        scheduler = ExpAnnealLR(optimizer, end_learning_rate, iterations)

        model_state = self.model.state_dict()
        self.model.train()

        step = 0
        while step < iterations:
            batch_loader = DataLoader(train_data, batch_size=mini_batch_size, shuffle=True)
            for batch in batch_loader:
                step += 1

                # forward pass
                loss = self.model.forward_loss(batch)
                if isinstance(loss, tuple):
                    loss = loss[0]

                # update optimizer and scheduler
                optimizer.zero_grad()
                loss.backward()
                torch.nn.utils.clip_grad_norm_(self.model.parameters(), 5.0)
                optimizer.step()
                scheduler.step(step)

                learning_rate = scheduler.get_lr()[0]

                loss_item = loss.item()
                if step == 1:
                    best_loss = loss_item
                else:
                    if smoothing_factor > 0:
                        moving_avg_loss = smoothing_factor * moving_avg_loss + (1 - smoothing_factor) * loss_item
                        loss_item = moving_avg_loss / (1 - smoothing_factor ** (step + 1))
                    if loss_item < best_loss:
                        best_loss = loss

                if step > iterations:
                    break

                if stop_early and (loss_item > 4 * best_loss or torch.isnan(loss)):
                    log_line(log)
                    log.info("loss diverged - stopping early!")
                    step = iterations
                    break

                with open(str(learning_rate_tsv), "a") as f:
                    f.write(f"{step}\t{datetime.datetime.now():%H:%M:%S}\t{learning_rate}\t{loss_item}\n")

            self.model.load_state_dict(model_state)
            self.model.to(flair.device)

        log_line(log)
        log.info(f"learning rate finder finished - plot {learning_rate_tsv}")
        log_line(log)

        return Path(learning_rate_tsv)
```

**Review by alanakbik — CHANGES_REQUESTED** [review]
@tadejmagajna Thanks for bringing this back! But some issues regarding the examples, see comments.

File: flair/hyperparameter/param_selection.py (hunk 1 above)
Comment (tadejmagajna) [reply flair/hyperparameter/param_selection.py]: As per your suggestion (which makes perfects sense - it felt really weird readding this outdated RNN code) I refactored `TextClassifierParamSelector` so that it uses `TransformerDocumentEmbeddings` now. I documented it to a reasonable extent in the tutorial.

File: flair/hyperparameter/parameter.py (hunk 2 above)
Comment (tadejmagajna) [reply flair/hyperparameter/parameter.py]: Did just that and refactored `TextClassifierParamSelector` so that it uses `TransformerDocumentEmbeddings`. Also renamed `DOCUMENT_EMBEDDING_PARAMETERS ` to `TEXT_CLASSIFICATION_PARAMETERS`

File: resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md (hunk 3 above)
Comment (tadejmagajna) [reply resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md]: I agree - makes sense to give precedence to sequence labelling in Flair.

As per your suggestion I added an example to the top that uses `WNUT_17 ` and documents the use of `SequenceTaggerParamSelector `

However, I also kept the example that shows how to use `TextClassifierParamSelector`. I feel like that if a feature isn't documented, people are unlikely to dig through the source and the value of the feature will be diminished.

So what I did was I created two sections (`Selecting hyperparameters for sequence labelling` and `Selecting hyperparameters for text classification`) and documented both use-cases accordingly.

File: resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md (hunk 5 above)
Comment (tadejmagajna) [reply resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md:194]: I addressed this as part of this PR, but note that I made tiny adjustment to the source code you pasted above. With the above code, we were getting the following warining from Torch:
> UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler.

this warning broke the CI build.

So I replaced the line `scheduler.step(step)` with  `scheduler.step()`. If my understanding is correct, this shouldn't badly affect the execution of the code. But if it does, please do let me know.

**Comment by tadejmagajna** [comment]
@alanakbik Thank you for the feedback!

I addressed your PR comments and I incorporated all of the suggestions including the refactoring of `TextClassifierParamSelector` so that it uses `TransformerDocumentEmbeddings` (instead of RNN) as well as fixing of the `find_learning_rate()`. I left more detailed comments in each individual PR feedback thread.

**Comment by tadejmagajna** [comment]
@alanakbik I made `fune_tune` a normal (non-tunable) param now and set it to default to True. Updated all the docs as well

**Review by alanakbik — APPROVED** [review]

**Comment by alanakbik** [comment]
@tadejmagajna thanks a lot - looks great!
